platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/HttpSolrClientFactorySpec.groovy

```diff
@@ -0,0 +1,185 @@
+      'not defined' || null          | SOLR_HOST2  | SOLR_PORT   | SOLR_CONTEXT   || SOLR_URL2 | CORE_URL2
+      'blank'       || ''            | SOLR_HOST2  | SOLR_PORT   | SOLR_CONTEXT   || SOLR_URL2 | CORE_URL2
+  }
+
+  @Timeout(HttpSolrClientFactorySpec.AVAILABLE_TIMEOUT_IN_SECS)
+  @Unroll
+  def 'test new client becoming available when system property solr.data.dir is #data_dir_is'() {
+    given:
+      def httpClient = Mock(SolrClient) {
+        ping() >> Mock(SolrPingResponse) {
+          // verify the Http Solr client should have been successfully pinged at least once
+          // must be done in 'given' because it will be called from a different thread and if
+          // declared in 'then', it will be out of scope and not matched
+          (1.._) * getResponse() >> Mock(NamedList) {
+            get("status") >> "OK"
+          }
+        }
+      }
+      def factory = Spy(HttpSolrClientFactory) {
+        // verify an actual Http Solr client will be created
+        // must be done in 'given' because it will be called from a different thread and if
+        // declared in 'then', it will be out of scope and not matched
+        1 * createSolrHttpClient(SOLR_URL, CORE, CORE_URL) >> httpClient
+      }
+
+    and:
+      System.setProperty("solr.http.url", SOLR_URL)
+
+    and:
+      System.setPropertyIfNotNull("solr.data.dir", solr_data_dir)
+
+    when:
+      def client = factory.newClient(CORE)
+
+    then: "the reported core should correspond to the requested one"
+      client.core == CORE
+
+    when: "checking for the adapter to become available"
+      // because the HttpSolrClientFactory wraps what it returns with a SolrClientAdapter which uses
+      // background threads to create and ping the server, we are forced to wait for it to become
+      // available. That being said, we mocked the client that will be internally created and its
+      // ping will be successful right away. Therefore this should all happen very quickly without
+      // a hitch. Worst case scenario we have a bug and it will take about 30 seconds to detect it.
+      // Best case scenario, it will return pretty much right away.
+      def available = client.isAvailable(AVAILABLE_TIMEOUT_IN_SECS + 5L, SECONDS)
+
+    then: "it should be"
+      available
+
+    and: "the underlying client should never be closed"
+      0 * httpClient.close()
+
+    and: "the config store is initialized and its data directory was or wasn't updated"
+      ConfigurationStore.instance.dataDirectoryPath == data_dir
+
+    where:
+      data_dir_is   || solr_data_dir || data_dir
+      'defined'     || DATA_DIR      || DATA_DIR
+      'not defined' || null          || null
+  }
+
+  def 'test new client with a null core'() {
+    given:
+      def factory = new HttpSolrClientFactory();
+
+    when:
+      factory.newClient(null)
+
+    then:
+      def e = thrown(IllegalArgumentException)
+
+      e.message.contains("invalid null Solr core")
```

Comment:
❓Do we really want to fail the test if the exception message is changed? This will only show up in logs, not on a UI so this seems a little too strict imho.

Reply:
the problem I have with blindly just checking for an IllegalArgumentException is that it could e coming from anything related or not. This way, I know for sure it is coming out for that reason.

Reply:
But then the test fails just because someone made a minor change (e.g., capitalized `Invalid`) to the message...

Reply:
Maybe so, but since it is a unit test, it should be caught right away